[spirv] Generate debug info for cbuffer

jaebaek opened the pull request:
This commit generates a `DebugGlobalVariable` instruction for a cbuffer
and creates a `DebugTypeComposite` for its type similar to struct/class.
Essentially we consider a cbuffer is a composite type global variable.

jaebaek committed:
[spirv] Generate debug info for cbuffer

This commit generates a `DebugGlobalVariable` instruction for a cbuffer
and creates a `DebugTypeComposite` for its type similar to struct/class.
Essentially we consider a cbuffer is a composite type global variable.

ehsannas commented on tools/clang/test/CodeGenSPIRV/rich.debug.cbuffer.hlsl:
```diff
@@ -0,0 +1,58 @@
+// Run: %dxc -T vs_6_0 -E main -fspv-debug=rich
+
+// CHECK:      OpName %type_MyCbuffer "type.MyCbuffer"
```
do you need to check for OpName and OpMemberName in this test? they are not related to your code.

jaebaek replied:
Removed.

ehsannas commented on tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:
```diff
@@ -894,6 +894,21 @@ SpirvVariable *DeclResultIdMapper::createCTBuffer(const HLSLBufferDecl *decl) {
+      const auto *varDecl = cast<VarDecl>(subDecl);
+      assert(varDecl && "Sub-Decls of HLSLBufferDecl must be VarDecls");
+
+      fields.emplace_back(varDecl->getType(), varDecl->getName());
```
I don't see `fields` being used anywhere here. What is the point of this loop?

jaebaek replied:
removed

ehsannas commented on tools/clang/lib/SPIRV/LowerTypeVisitor.cpp:
```diff
@@ -87,6 +87,15 @@ bool LowerTypeVisitor::visitInstruction(SpirvInstruction *instr) {
+      auto *actualType = varType->getPointeeType();
+      debugGlobalVar->setDebugSpirvType(actualType);
+      if (auto *structType = dyn_cast<StructType>(actualType))
+        lowerDebugTypeComposite(structType, structType->getFields(), false,
```
Why do you need to pass `structType->getFields()` to the function?

`lowerDebugTypeComposite` has access to `structType` and can call `structType->getFields()` itself.

jaebaek replied:
done

ehsannas commented on tools/clang/lib/SPIRV/LowerTypeVisitor.h:
```diff
@@ -98,6 +98,10 @@ class LowerTypeVisitor : public Visitor {
       const RecordType *structType, const SpirvType *spirvType,
       llvm::SmallVector<StructType::FieldInfo, 4> &fields, bool isResourceType);
+  SpirvDebugTypeComposite *
+  lowerDebugTypeComposite(const SpirvType *type,
```
What is the difference between this function and the one on line 98?

ehsannas replied:
The Google C++ guide has some info about [function overloading](https://google.github.io/styleguide/cppguide.html#Function_Overloading). Here we have two functions with the same name without clear indication why one should be used and not the other. If this one is only used for cbuffers for example, you can change the name to indicate that.

jaebaek replied:
Line 98 was for `QualType` based composite type lowering and this is for `SpirvType` based composite type lowering.

For some cases like `cbuffer`, we do not have `QualType` for it like a `HybridType`.
I changed the function name and add a comment.

ehsannas replied:
So Spiregg goes through the AST, for a struct we either have a `QualType` or a `HybridStructType`.

so why is `lowerDebugTypeCompositeFromSpirvType` taking a `StructType` rather than a `HybridStructType`?

ehsannas commented on tools/clang/lib/SPIRV/LowerTypeVisitor.cpp:
```diff
@@ -972,6 +981,66 @@ LowerTypeVisitor::generateFunctionInfo(const CXXMethodDecl *decl,
   return fn;
 }
 
+SpirvDebugTypeComposite *LowerTypeVisitor::lowerDebugTypeComposite(
```
It looks like this function has a lot of duplicate code from the other `lowerDebugTypeComposite` function.

jaebaek replied:
Right. Let me consider how to refactor it.

jaebaek committed:
Update based on code review

jaebaek reviewed (commented):
Thank you for your code review. It helps a lot.
I did not finish refactoring `LowerTypeVisitor::lowerDebugTypeComposite()` yet. Will add one more commit.

ehsannas reviewed (commented):
The whole way we're doing lowering for debug composites is very confusing to me. Some part of it is done in LowerTypeVisitor, and some parts in DebugTypeVisitor.

Ideally all of it should be in DebugTypeVisitor. I know it may not be 100% possible. So we should try to make the LowerTypeVisitor part as small as possible, and make clear comments about why they are in LowerTypeVisitor instead of DebugTypeVisitor.

For example: it seems to me that `lowerDebugTypeCompositeFromSpirvType` can be moved to DebugTypeVisitor?

jaebaek committed:
Move LowerTypeVisitor::lowerDebugTypeCompositeFromSpirvType() to DebugTypeVisitor

If a debug instruction does not have QualType, we do not have to lower
it in LowerTypeVisitor. DebugTypeVisitor exists exactly for that case.

jaebaek commented:
I agree that handling the debug composite type is complicated.
I moved `lowerDebugTypeCompositeFromSpirvType` to DebugTypeVisitor as you recommended.

However, I do not have clear idea how to reduce the code complexity of handling the debug composite type for now .. The main issue is that the debug composite type requires more information than other debug types. It needs the type declaration and QualType information. I will open a bug to remind ourselves that we want to refactor it in a better way.

ehsannas commented on tools/clang/lib/SPIRV/DebugTypeVisitor.cpp:
```diff
@@ -260,12 +303,28 @@ bool DebugTypeVisitor::visitInstruction(SpirvInstruction *instr) {
+          spirvType = ptrType->getPointeeType();
+          debugGlobalVar->setDebugSpirvType(spirvType);
+          if (auto *structType = dyn_cast<StructType>(spirvType))
+            prepareDebugTypeComposite(structType, instr->getSourceLocation());
```
It would be better to call this `lowerCbufferDebugType` so that it doesn't get mixed up with regular structures

jaebaek replied:
done

jaebaek committed:
Refactoring DebugTypeVisitor

ehsannas reviewed (approved):
Nice! Thanks

ehsannas commented:
Oh no!

jaebaek commented:
> Oh no!

Hmm I think the failure is not related to this PR. I cannot reproduce the error on my local Windows/Linux machine.
I think this is a timing issue.

Could you please run this once more? (I cannot run it)

ehsannas commented:
I've triggered a rebuild

jaebaek commented:
Thanks!
